src/test/python/parser/opentrackio_lib.py

```diff
@@ -137,60 +234,66 @@ def get_timecode(self):
+            return self.pd["timing"]["synchronization"]["source"]
+        else:
+            return None
 
-    def get_sample_time(self,part=None):
-        """Get the PTP sample time in user-preferred format"""
-        if self.validate_dict_elements(self.pd,["timing","sampleTimestamp","seconds"]):
+    def get_sample_time(self, format: Optional[TimeFormat] = None, part=None):
+        """Get the PTP/NTP sample time in user-preferred format."""
+        if self.validate_dict_elements(self.pd, ["timing", "sampleTimestamp", "seconds"]):
             ssec = int(self.pd["timing"]["sampleTimestamp"]["seconds"])
             nsec = int(self.pd["timing"]["sampleTimestamp"]["nanoseconds"])
-            # Constants
-            epoch = 1970                        # PTP is since this epoch
-            spm = 60                            # seconds per minute. Common knowledge, but consistency is important
-            sph = 3600                          # seconds per hour
-            spd = 86400                         # sec per day
-            spy = 31536000                      # sec per year
-            # separate into years, days, hours, min, sec
-            ydelta = int(ssec / spy)            # years since epoch
-            yr = int(epoch + ydelta)            # current year
-            sty = int(ssec - (ydelta * spy))    # seconds elapsed this year
-            day = int(sty / spd)                # current day of year
-            std = int(sty - (day * spd))        # seconds elapsed today (since midnight)
-            hr = int(std / sph)                 # hours elapsed today
-            mn = int((std - (hr * sph)) / spm)  # remainder minutes
-            st = int(std - (hr * sph) - (mn * spm))  # remainder seconds
-        if not part:
-            if self.sample_time_format == "sec":
-                return ssec + (nsec * 0.000000001)
-            elif self.sample_time_format == "timecode":             # since midnight
-                frm = int((nsec * 0.000000001) * self.get_timecode_framerate())
-                return '{:02}'.format(hr) + ":" + '{:02}'.format(mn) + ":" + '{:02}'.format(st) + ":" + '{:02}'.format(frm)
-            elif self.sample_time_format == "string":
-                return "year:{} day:{} hour:{} min:{} sec:{} nsec:{}".format(yr,day,hr,mn,st,nsec)
-        else:
-            if part == 'yy':
-                return yr
-            elif part == 'dd':
-                return day
-            elif part == 'hh':
-                return hr
-            elif part == 'mm':
-                return mn
-            elif part == 'ss':
-                return st
-            elif part == 'ns':
-                return nsec
+        
+            # Calculate the total time as seconds plus fractional seconds from nanoseconds and attoseconds
+            total_time = ssec + (nsec * 1e-9)
+            base_time = datetime(1970, 1, 1) + timedelta(seconds=ssec, microseconds=nsec / 1_000)
```

Comment:
This new approach is library-based. The original code was written as non-Pythonic, non library-based, and as simple as possible, in order to facilitate porting to other less-capable embedded languages like C.  Making the code more Pythonic and library-based makes the code less easy to port. But if you don't think that is a priority, I can live with it this way. My primary goal in writing the parser was to have /something/ in place for the first public release and big announcement at MTS.